how-it-works/1_protocol/03_message-routing.subpage.md

```diff
@@ -7,8 +7,128 @@ slug: message-routing
+
+## Inter-Canister Messaging
+
+As mentioned above, the execution of a canister message can lead to the creation of a new inter-canister message sent to a local or remote (on a different subnet) canister.
+Let us go deeper into how inter-canister messaging works.
+
+### Intra-Subnet Inter-Canister Messaging
+
+Intra-subnet, i.e., local inter-messages originating from an executing canister method do not need to go through consensus as they deterministically result from messages that have been agreed by a previous consensus round and their further execution remains completely deterministic.
+This holds transitively, that is, inter-canister messages can create new inter-canister messages, resulting in a tree of messages.
+Local message invocations can be executed as long as the cycles limit for the round has not yet been exhausted.
+It is important to note that this local inter-canister messaging is *not synchronous* message invocation as you might be used from EVM-based blockchains.
+Rather, local messages are put into the input queue of the target canister and are scheduled for execution asynchronously.
+This is the standard inter-canister messaging semantics known for the Internet Computer.
+
+### Inter-Subnet Inter-Canister Messaging
+
+Remote inter-canister messages, that is, such sent to canisters on other subnets, are handled differently.
+A remote message is placed into the outgoing *subnet queue* for the target subnet once it results from the execution of a canister message.
+The xnet messages in the queue are certified (signed) using a Merkle-tree-style data representation, at the end of the round by the subnet using BLS threshold cryptography as part of the per-round state certification.
+That is, every message in the outgoing queue is certified by the originating subnet.
+Replicas on the receiving subnet obtain the xnet message during block making (part of consensus), validate the threshold signature, and include a valid xnet message in a consensus block.
+Thanks to using a Merkle-tree-like xnet queue, parts of the queue can be consumed in a round by the receiving subnets and signatures can still be validated.
+
+## State Certification
+
+The replicated state of a subnet comprises all the relevant information required for the operation of the subnet:
+* Items certified per round:
+  * Responses to ingress messages
+  * Xnet messages to be sent to other subnets
+  * Canister metadata (module hashes, *certified variables*)
+* Items certified per checkpoint:
+  * All canister Wasm bytecode
+  * All canister state (heap and stable memory)
 
+The message routing layer performs a certification of parts of the replicated state after each round (the first three items above) and certification of the full replicated state (including also the remaining two items above) at every checkpoint, which is typically every multiple-hundred rounds.
```

Comment:
I think this sentence should either move before the enumeration above, or be dropped entirely as it just repeats what is already said in the enumeration.

Reply:
```suggestion
```